[ML] Adding v3 modules for Security_Linux and Security_Windows and Deprecating v1 + v2

bfilar opened the pull request:
## Summary
- Follow up to a closed PR in 8.1 https://github.com/elastic/kibana/pull/123274 
- Follow up to another closed PR in 8.2 https://github.com/elastic/kibana/pull/128099
- This PR seeks to consolidate the current set of Security ML Modules: https://github.com/elastic/security-team/issues/2960
- Added files for new + refactored v3 jobs for both `security_linux` and `security_windows` - for use within the Security app. 
- Deprecated v1 and v2 versions of those jobs
- Deprecated Auditbeat Host Process module due to a duplicative job in the Linux Security module
- Consolidated auth modules for Linux and Windows will be consolidated into their respective OS modules

### Files/Job Artifacts:
---
- 2 updated manifest `.json` files - for both linux and windows

- Updated/new ML Job configurations for 26 jobs - each with associated datafeed configuration files:
  - **security_linux: 14 jobs**
    - v3_linux_anomalous_network_activity
    - v3_linux_anomalous_network_port_activity_ecs
    - v3_linux_anomalous_process_all_hosts_ecs
    - v3_linux_anomalous_user_name_ecs
    - v3_linux_network_configuration_discovery
    - v3_linux_network_connection_discovery
    - v3_linux_rare_metadata_process
    - v3_linux_rare_metadata_user
    - v3_linux_rare_sudo_user
    - v3_linux_rare_user_compiler
    - v3_linux_system_information_discovery
    - v3_linux_system_process_discovery
    - v3_linux_system_user_discovery
    - v3_rare_process_by_host_linux_ecs
    
  - **security_windows: 12 jobs**
    - v3_rare_process_by_host_windows_ecs
    - v3_windows_anomalous_network_activity_ecs
    - v3_windows_anomalous_path_activity_ecs
    - v3_windows_anomalous_process_all_hosts_ecs
    - v3_windows_anomalous_process_creation
    - v3_windows_anomalous_script
    - v3_windows_anomalous_service
    - v3_windows_anomalous_user_name_ecs
    - v3_windows_rare_metadata_process
    - v3_windows_rare_metadata_user
    - v3_windows_rare_user_runas_event
    - v3_windows_rare_user_type10_remote_login

### Tests:
---
Individual job test tracking stats available here: https://docs.google.com/spreadsheets/d/1JOUIVsitaMdEdhM3WT2Eag4ELI-rI2Jec7bXildJsdQ/edit#gid=0

@randomuserid to also post more updates as needed to this issue + regarding tests, thanks

bfilar committed:
consolidate Security ML Modules

bfilar committed:
removal of auditbeat host processes ecs module

bfilar committed:
removing siem_winlogbeat_auth after consolidating into windows_security

bfilar committed:
renamed to avoid job collisions

randomuserid committed:
Update recognize_module.ts

removed references to deprecated v1 modules which no longer exist

randomuserid committed:
test fixes

remove references to deprecated module and modify module names to match the latest v3 modules being committed.

kibanamachine committed:
Merge branch 'main' into bfilar-ml-refactor

randomuserid committed:
Update recognize_module.ts

think this is what the linter wants

randomuserid committed:
Merge branch 'bfilar-ml-refactor' of https://github.com/elastic/kibana into bfilar-ml-refactor

kibanamachine committed:
Merge branch 'main' into bfilar-ml-refactor

bfilar committed:
deprecating winlogbeat and auditbeat modules

bfilar committed:
fixes test post-deprecation of modules

bfilar committed:
fixes typo in test

bfilar committed:
revert linting changes

bfilar committed:
revert linting changes pt2

bfilar committed:
fixing test in setup_module.ts

bfilar committed:
ml module refactor

bfilar commented:
@peteharverson @pheyos, the SecML team will coordinate w/ @spong to identify the engineering stakeholders for failing tests. We will seek to resolve these in the coming weeks.

spong commented:
> @peteharverson @pheyos, the SecML team will coordinate w/ @spong to identify the engineering stakeholders for failing tests. We will seek to resolve these in the coming weeks.

Sounds good! We have this issue for tracking the rework of these tests in support of the v3 modules: https://github.com/elastic/kibana/issues/128318

Ideally we'd like to coordinate that rework as part of this PR to ensure coverage, but if they need to happen as a follow-up as to not block you folks that's fine as well too. So feel free to skip any Security Solution tests (maybe add a `TODO: PR 131166` comment above them) that are giving you trouble for the time being.

randomuserid commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into bfilar.ml-refactor-2

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_linux/ml/v3_linux_anomalous_network_activity.json:
```diff
@@ -0,0 +1,54 @@
+    "time_field": "@timestamp"
+  },
+  "custom_settings": {
+    "custom_urls": [
```
Does it matter that these URLs use `siem#` rather than starting with `security`?

bfilar replied:
@randomuserid any thoughts/preferences here?

randomuserid replied:
they still work in main. Only the UI labels were renamed from siem to security, not everything in the backend.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/manifest.json:
```diff
@@ -30,84 +30,119 @@
     {
-      "id": "v2_windows_anomalous_process_all_hosts_ecs",
-      "file": "v2_windows_anomalous_process_all_hosts_ecs.json"
+      "id": "v3_rare_process_by_host_windows_ecs",
```
As with the `security_linux` jobs, I'd recommend removing the `_ecs` suffix from all job and datafeed IDs and filenames. Unless there is a specific reason why you'd want to retain it for only certain jobs in the module?

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_rare_process_by_host_windows_ecs.json:
```diff
@@ -0,0 +1,47 @@
+{
+    "job_id": "v3_rare_process_by_host_windows_ecs",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_windows_anomalous_network_activity_ecs.json:
```diff
@@ -0,0 +1,71 @@
+{
+    "job_id": "v3_windows_anomalous_network_activity_ecs",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_windows_anomalous_path_activity_ecs.json:
```diff
@@ -0,0 +1,47 @@
+{
+    "job_id": "v3_windows_anomalous_path_activity_ecs",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_windows_anomalous_process_all_hosts_ecs.json:
```diff
@@ -0,0 +1,47 @@
+{
+    "job_id": "v3_windows_anomalous_process_all_hosts_ecs",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_windows_anomalous_process_creation.json:
```diff
@@ -0,0 +1,47 @@
+{
+    "job_id": "v3_windows_anomalous_process_creation",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_windows_anomalous_script.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "job_id": "JOB_ID",
+  "job_id": "v3_windows_anomalous_script",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_windows_anomalous_service.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "job_id": "JOB_ID",
+  "job_id": "v3_windows_anomalous_service",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_windows_anomalous_user_name_ecs.json:
```diff
@@ -0,0 +1,47 @@
+{
+    "job_id": "v3_windows_anomalous_user_name_ecs",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_windows_rare_metadata_process.json:
```diff
@@ -0,0 +1,23 @@
+{
+    "job_id": "v3_windows_rare_metadata_process",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_windows_rare_metadata_user.json:
```diff
@@ -0,0 +1,23 @@
+{
+    "job_id": "v3_windows_rare_metadata_user",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_windows_rare_user_runas_event.json:
```diff
@@ -0,0 +1,42 @@
+{
+    "job_id": "v3_windows_rare_user_runas_event",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/datafeed_v3_windows_rare_user_type10_remote_login.json:
```diff
@@ -1,11 +1,11 @@
-  "max_empty_searches": 10,
-  "query": {
-    "bool": {
+    "job_id": "v3_windows_rare_user_type10_remote_login",
```
As above, this should be the placeholder `JOB_ID`.

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/v3_windows_anomalous_script.json:
```diff
@@ -0,0 +1,46 @@
+    "time_field": "@timestamp"
+  },
+  "custom_settings": {
+    "custom_urls": [
```
Does it matter that these URLs use `siem#` rather than starting with `security` ?

peteharverson reviewed (commented):
Taken a first look at these jobs, and left some comments on the configs. 

I was able to verify that the jobs successfully show up in the ML job wizard and are created without error, but the data sets I tried all seem fairly out of date for the queries used in the datafeeds, so will leave approval of the jobs themselves to you @randomuserid

bfilar committed:
manifest, job, and datafeed cleanup based on PR feedback

bfilar committed:
Merge branch 'bfilar.ml-refactor-2' of github.com:elastic/kibana into bfilar.ml-refactor-2

bfilar committed:
commenting out security solution tests for ML Modules

lcawl commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/v3_windows_anomalous_script.json:
```diff
@@ -0,0 +1,46 @@
+    "bucket_span": "15m",
+    "detectors": [
+      {
+        "detector_description": "high_info_content(\"powershell.file.script_block_text\")",
```
```suggestion
        "detector_description": "Detects high information content in powershell.file.script_block_text values.)",
```

lcawl commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/v3_windows_anomalous_user_name.json:
```diff
@@ -1,22 +1,23 @@
     "bucket_span": "15m",
     "detectors": [
       {
         "detector_description": "rare by \"user.name\"",
```
```suggestion
        "detector_description": "Detects rare user.name values.",
```

lcawl commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/v3_windows_rare_metadata_process.json:
```diff
@@ -16,7 +15,8 @@
       {
         "detector_description": "rare by \"process.name\"",
```
```suggestion
        "detector_description": "Detects rare process.name values.",
```

lcawl commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/v3_windows_rare_metadata_user.json:
```diff
@@ -16,7 +15,8 @@
       {
         "detector_description": "rare by \"user.name\"",
```
```suggestion
        "detector_description": "Detects rare user.name values.",
```

lcawl reviewed (commented):
I've added suggestions for the detector descriptions to align with what we've done in the job wizards

bfilar commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_linux/ml/v3_linux_anomalous_network_activity.json:
```diff
@@ -0,0 +1,54 @@
+    "bucket_span": "15m",
+    "detectors": [
+      {
+        "detector_description": "rare by \"process.name\"",
```
Thank you @lcawl really appreciate the feedback

randomuserid commented:
> Taken a first look at these jobs, and left some comments on the configs.
> 
> I was able to verify that the jobs successfully show up in the ML job wizard and are created without error, but the data sets I tried all seem fairly out of date for the queries used in the datafeeds, so will leave approval of the jobs themselves to you @randomuserid

Yes that is expected unless the test data is very recent. The index names, field names, and queries have changed several times between versions 1-3 of these modules.

bfilar committed:
modified ml module tests and job descriptions

randomuserid commented:
@elasticmachine merge upstream

randomuserid committed:
Merge branch 'main' into bfilar.ml-refactor-2

kibanamachine committed:
Merge branch 'main' into bfilar.ml-refactor-2

peteharverson commented on x-pack/test/api_integration/apis/ml/modules/setup_module.ts:
```diff
@@ -311,33 +311,6 @@ export default ({ getService }: FtrProviderContext) => {
       },
     },
-    {
-      testTitleSuffix:
```
Noting that removal of these tests leaves us with no CI coverage inside ML (API integration or functional tests) for the creation of the `security_linux` and `security_windows` jobs. The ML data team are looking at creating some new downsampled data sets which we can use in CI, and we will aim to add back some tests in a follow-up, so no need to hold up this PR.

randomuserid committed:
Update datafeed_auth_high_count_logon_events_for_a_source_ip.json

added test for existence of source.ip field per https://github.com/elastic/kibana/issues/131376

randomuserid committed:
Update datafeed_auth_high_count_logon_events_for_a_source_ip.json

formatting

randomuserid committed:
Merge branch 'main' into bfilar.ml-refactor-2

randomuserid committed:
descriptions

standardized descriptions between Linux and Windows jobs; removed the term "services" from the rare process jobs because it has a special meaning under Windows and is the target of a different job; added a sentence to the sudo job description, I think this was a stub description that never got fleshed out when it was developed.

randomuserid committed:
tags

added job tags

randomuserid committed:
tags

added Linux job tags

randomuserid committed:
tags

randomuserid reviewed (approved):
Good, Linux & Windows modules have been refactored and the v1/v2 modules removed. I checked all of the memory sizes and descriptions. I also added the job tags.

randomuserid committed:
Merge branch 'main' into bfilar.ml-refactor-2

randomuserid committed:
Merge branch 'main' into bfilar.ml-refactor-2

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/v3_windows_anomalous_script.json:
```diff
@@ -0,0 +1,53 @@
+{
+  "job_type": "anomaly_detector",
+  "description": "Looks for unusual powershell scripts that may indicate execution of malware, or persistence mechanisms.",
```
To be consistent with all the other descriptions in this module, should add `Security: Windows - ` onto the front of the text here.

randomuserid replied:
done

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_linux/ml/v3_linux_anomalous_network_activity.json:
```diff
@@ -0,0 +1,62 @@
+  "data_description": {
+    "time_field": "@timestamp"
+  },
+  "custom_settings": {
```
The `custom_settings` for all the `security_linux` jobs is malformed - they contain an extra` custom_settings` level inside `custom_settings`, and the JSON is malformed, which prevents the jobs loading correctly in the ML job wizard. I see this error in the Kibana server console:

```
[2022-05-17T15:19:05.344+01:00][WARN ][plugins.ml] Data recognizer error loading config for job v3_linux_anomalous_network_port_activity for module security_linux_v3. SyntaxError: Unexpected end of JSON input
[2022-05-17T15:19:05.344+01:00][WARN ][plugins.ml] Data recognizer error loading config for job v3_linux_network_configuration_discovery for module security_linux_v3. SyntaxError: Unexpected end of JSON input
[2022-05-17T15:19:05.345+01:00][WARN ][plugins.ml] Data recognizer error loading config for job v3_linux_network_connection_discovery for module security_linux_v3. SyntaxError: Unexpected end of JSON input
```

randomuserid replied:
done

peteharverson reviewed (changes requested):
The JSON for the `security_linux` jobs is going to need fixing before I can test these inside the ML job wizards.

randomuserid committed:
linting

remove a dup json element

randomuserid committed:
Merge branch 'bfilar.ml-refactor-2' of https://github.com/elastic/kibana into bfilar.ml-refactor-2

randomuserid committed:
Update v3_windows_anomalous_script.json

add the Security: Windows prefix which was missing

randomuserid committed:
Update v3_linux_anomalous_network_activity.json

missing bracket

randomuserid committed:
Merge branch 'main' into bfilar.ml-refactor-2

peteharverson commented on x-pack/plugins/ml/server/models/data_recognizer/modules/security_windows/ml/v3_windows_anomalous_script.json:
```diff
@@ -0,0 +1,53 @@
+    "bucket_span": "15m",
+    "detectors": [
+      {
+        "detector_description": "Security: Windows - Detects high information content in powershell.file.script_block_text values.",
```
You made the edit to the `detector_description`, but it should be in the job `description` field above. The convention used for all the other jobs is to start all the job descriptions with `Security: Windows - `

randomuserid replied:
fixed

randomuserid committed:
Update v3_windows_anomalous_script.json

the prefix was in the wrong place

peteharverson reviewed (approved):
Tested job creation works inside the ML job wizard. LGTM.

lcawl reviewed (approved):
Text strings LGTM

spong reviewed (approved):
Checked out, tested locally, and verified functionality of the new `v3 modules` within the Security Solution app. Was able to install and start a job from the `ML Job Settings` UI, and then created a detection rule referencing the job without issue.

Note I: I did see one warning in the Kibana console after initiating the install, which hits the `/api/ml/modules/setup/security_windows_v3` API with this payload:

##### Payload
<img width="956" alt="image" src="https://user-images.githubusercontent.com/2946766/168909313-f2bbda17-3e78-42a5-b021-eaec3b3b42e4.png">

##### Kibana Console Warning
> [2022-05-17T14:50:44.805-06:00][WARN ][plugins.ml] Data recognizer could not estimate model memory limit {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"[Overall] cardinality estimate required for [by_field_name] [winlog.event_data.ServiceName] but not supplied"}],"type":"illegal_argument_exception","reason":"[Overall] cardinality estimate required for [by_field_name] [winlog.event_data.ServiceName] but not supplied"},"status":400}

@peteharverson, is there something we need to update on our end with this request, or is this expected?

 ---

Note II: With regards to the skipped tests, we have this issue for tracking https://github.com/elastic/kibana/issues/128318 and @banderror will be following up with a PR to unskip these tests and move them to the new `v3 modules`.

All that said, LGTM from the Security Solution side of the house! 😀 Excited for the consolidation here -- thanks @bfilar & @randomuserid! <img width="16" src="https://user-images.githubusercontent.com/2946766/141219243-86619f7f-c300-4355-96f5-316d70aa6a0e.png" />

yctercero reviewed (approved):
Just looking at the exceptions cypress test (code owners) - if we can track that skip that'd be great.

dhurley14 reviewed (approved):
LGTM! Looks @elastic/security-solution-platform only needed to review a skipped test.

peteharverson commented:
> is there something we need to update on our end with this request, or is this expected?

```
[2022-05-17T14:50:44.805-06:00][WARN ][plugins.ml] Data recognizer could not estimate model memory limit {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"[Overall] cardinality estimate required for [by_field_name] [winlog.event_data.ServiceName] but not supplied"}],"type":"illegal_argument_exception","reason":"[Overall] cardinality estimate required for [by_field_name] [winlog.event_data.ServiceName] but not supplied"},"status":400}
```

@spong in the data set you are using for your tests, are there any docs with `winlog.event_data.ServiceName` used as the `by_field_name` for the `v3_windows_anomalous_service` job? The checks used to estimate the model memory limit for the job will not succeed if they cannot obtain an estimate for the cardinality of some of the fields used in the job. I get the same error using my test data set, where none of the docs contain `winlog.event_data.ServiceName`.

spong commented:
> @spong in the data set you are using for your tests, are there any docs with `winlog.event_data.ServiceName` used as the `by_field_name` for the `v3_windows_anomalous_service` job? The checks used to estimate the model memory limit for the job will not succeed if they cannot obtain an estimate for the cardinality of some of the fields used in the job. I get the same error using my test data set, where none of the docs contain `winlog.event_data.ServiceName`.

Ahh, that's indeed the case, my sample data does not include those fields either. Thanks for confirming Pete! 🙂
